### PR TITLE
[SPARK-48816][SQL] Shorthand for interval converters in UnivocityParser

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Directly call `IntervalUtils.castStringToDTInterval/castStringToYMInterval` instead of creating Cast expressions to evaluate. 

- Benchmarks indicated a 10% time-saving.
- Bad record recording might not work if the cast handles the exceptions early

### Why are the changes needed?

- pref improvement
- Bugfix for bad record recording

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

passing existing tests and benchmark tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmark.scala [hunk 1]
```diff
@@ -371,6 +371,38 @@ object CSVBenchmark extends SqlBasedBenchmark {
     }
   }
 
+  private def intervalBenchmark(rowsNum: Int, numIters: Int): Unit = {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmark.scala:374]: Thank you for adding this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmark.scala [hunk 2]
```diff
@@ -379,6 +411,7 @@ object CSVBenchmark extends SqlBasedBenchmark {
       countBenchmark(rowsNum = 10 * 1000 * 1000, numIters)
       datetimeBenchmark(rowsNum = 10 * 1000 * 1000, numIters)
       filtersPushdownBenchmark(rowsNum = 100 * 1000, numIters)
+      intervalBenchmark(rowsNum = 300_1000, numIters)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVBenchmark.scala:414]: Sorry but shall we avoid this new syntax pattern, `300_1000`, from Spark repo, @yaooqinn ? :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 3]
```diff
@@ -260,12 +260,14 @@ class UnivocityParser(
 
     case ym: YearMonthIntervalType => (d: String) =>
       nullSafeDatum(d, name, nullable, options) { datum =>
-        Cast(Literal(datum), ym).eval(EmptyRow)
+        IntervalUtils.castStringToYMInterval(
+          UTF8String.fromString(datum), ym.startField, ym.endField)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:264]: From the PR description, is this a behavior change?
> Bad record recording might not work if the cast handles the exceptions early

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:264]: Not really, users still get NULL after this

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side. Thank you. Feel free to merge, @yaooqinn .

**Comment by yaooqinn** [comment]
Merged to master. Thank you again @dongjoon-hyun

**Pull request closed without merging**